### PR TITLE
[oneDNN] sum op refactor

### PR DESCRIPTION
**Pull request by jczaja**
### PR types
Others

### PR changes
OPs

### Describe
This is reimplementation of sum oneDNN op to use AcquireAPI and that way to adapt it easily for bf16 support (next PR)

### CODE REVIEW

**Comment by jczaja** [comment]
@arogowie-intel  Please review

**Review by wozna — APPROVED** [review]
LGTM

File: paddle/fluid/operators/mkldnn/sum_mkldnn_op.cc [hunk 1]
```diff
@@ -59,85 +148,67 @@ class SumMKLDNNOpKernel : public paddle::framework::OpKernel<T> {
...
-    std::vector<mkldnn::memory> srcs_mem;
+    bool in_place = (input0.numel() > 0) && input0.IsSharedBufferWith(*output);
 
-    auto& input0 = in_vars[0]->Get<LoDTensor>();
-    in_place = (input0.numel() > 0) && (input0.data<T>() == output_data);
+    SumMKLDNNHandler<T> handler(dev_ctx, ctx.GetPlace(), in_vars, output,
+                                ctx.OutputName("Out"));
 
+    // Create list of SRC MEMs
+    std::vector<std::shared_ptr<mkldnn::memory>> srcs_mem;
```
Comment (arogowie-intel) [inline paddle/fluid/operators/mkldnn/sum_mkldnn_op.cc:164]: There is using at the beginning of file for `mkldnn::memory`. Please unify usage, or delete it.
```suggestion
    std::vector<std::shared_ptr<memory>> srcs_mem;
```

File: paddle/fluid/operators/mkldnn/sum_mkldnn_op.cc [hunk 2]
```diff
@@ -59,85 +148,67 @@ class SumMKLDNNOpKernel : public paddle::framework::OpKernel<T> {
...
+          output, in_out.format(), ctx.GetPlace());
+
+      auto reorder_p = reorder_handler.AcquireReorder(target_mem, dst_mem);
       reorder_p->execute(astream, *dst_mem, *target_mem);
       astream.wait();
     }
-
-    output->set_layout(DataLayout::kMKLDNN);
-    output->set_format(output_format);
+    output->set_layout(framework::DataLayout::kMKLDNN);
```
Comment (arogowie-intel) [inline paddle/fluid/operators/mkldnn/sum_mkldnn_op.cc:210]: There is using at the beginning of file for `framework::DataLayout`. Please unify usage, or delete it.

File: paddle/fluid/operators/mkldnn/sum_mkldnn_op.cc [hunk 3]
```diff
@@ -59,85 +148,67 @@ class SumMKLDNNOpKernel : public paddle::framework::OpKernel<T> {
...
-    mkldnn::stream astream(mkldnn_engine);
     std::unordered_map<int, memory> args;
     for (size_t i = 0; i < srcs_mem.size(); ++i) {
-      args.insert({MKLDNN_ARG_MULTIPLE_SRC + i, srcs_mem.at(i)});
+      args.insert({MKLDNN_ARG_MULTIPLE_SRC + i, *(srcs_mem[i])});
     }
     args.insert({MKLDNN_ARG_DST, *dst_mem});
 
-    sum_prim.execute(astream, args);
+    mkldnn::stream astream(dev_ctx.GetEngine());
```
Comment (arogowie-intel) [inline paddle/fluid/operators/mkldnn/sum_mkldnn_op.cc:187]: There is using for `mkldnn::stream` at the beginning of file. Please unify usage, or delete it.

File: paddle/fluid/operators/mkldnn/sum_mkldnn_op.cc [hunk 4]
```diff
@@ -51,6 +51,95 @@ using paddle::platform::CPUDeviceContext;
...
+  // we cannot use base AcquireForwardPrimitiveDescriptor
+  void AcquireForwardPrimitiveDescriptor(
+      const memory::desc& dst_md, const std::vector<float>& scales,
+      const std::vector<memory::desc>& srcs_md) {
+    // Sum op does not have backward so no passing from FWD to BWD is needed
+    const std::string key_pd = this->key_ + "@fwd_pd";
+    this->fwd_pd_ = std::static_pointer_cast<dnnl::sum::primitive_desc>(
+        this->dev_ctx_.GetBlob(key_pd));
+    if (this->fwd_pd_ == nullptr) {
+      this->fwd_pd_.reset(new mkldnn::sum::primitive_desc(
```
Comment (arogowie-intel) [inline paddle/fluid/operators/mkldnn/sum_mkldnn_op.cc:104]: There is using for `mkldnn::sum` at the beginning of file. Please unify usage, or delete it.

File: paddle/fluid/operators/mkldnn/sum_mkldnn_op.cc [hunk 5]
```diff
@@ -51,6 +51,95 @@ using paddle::platform::CPUDeviceContext;
...
+          dst_md, scales, srcs_md, this->engine_));
+      this->dev_ctx_.SetBlob(key_pd, this->fwd_pd_);
+    }
+  }
+
+  std::shared_ptr<mkldnn::memory> AcquireSrcMemory(
+      const framework::Tensor& input, int i) {
```
Comment (arogowie-intel) [inline paddle/fluid/operators/mkldnn/sum_mkldnn_op.cc:111]: There is using for `paddle::framework::Tensor` at the beginning of file. Please unify usage or, delete it.

**Review by arogowie-intel — APPROVED** [review]
Just few minor comments. I'd suggest to clean those multiple `using`s at the beginning of file. Otherwise LGTM.

**Comment by jczaja** [comment]
@arogowie-intel Many thanks for review. All suggesstions are of style/cosmetic fixes, so I will fix it in next PR (bf16 support to sum).

@luotao1 Could you please start your review?

File: python/paddle/fluid/tests/unittests/mkldnn/test_sum_mkldnn_op.py [hunk 6]
```diff
@@ -86,4 +86,6 @@ def test_check_grad(self):
 
 
 if __name__ == '__main__':
+    from paddle import enable_static
```
Comment (lidanqing-vv) [inline python/paddle/fluid/tests/unittests/mkldnn/test_sum_mkldnn_op.py:89]: Just a comment, not asking for changes. Baidu recently added STATIC_MODE_TESTING_LIST, python unit test do not require enable_static() any more.

Comment (jczaja) [reply python/paddle/fluid/tests/unittests/mkldnn/test_sum_mkldnn_op.py:89]: ok, will try to remove it in next PR

**Review by luotao1 — APPROVED** [review]

**Pull request merged**